Add "Fetcher" class for bulk fetching of shortcode previews

goldenapples opened the pull request:
Adds a `Fetcher` utility class to combine and bulk-send requests for
shortcode previews. Currently - until I can find some reliable events on
MCE processing to hook to - this is being done in a very naive way:
every time a request is made, it sets a half-second timeout, so that any
other requests needed can be gathered and sent together.

See #371

goldenapples committed:
Add "Fetcher" class for bulk fetching on shortcode views

Adds a Fetcher utility class to combine and bulk-send requests for
shortcode previews. Currently - until I can find some reliable events on
MCE processing to hook to - this is being done in a very naive way:
every time a request is made, it sets a half-second timeout, so that any
other requests needed can be gathered and sent together.

goldenapples committed:
Delete timeout flag after returning values

The `fetcher.timeout` flag is set when initiating a batch request, to
ensure that subsequent requests get added to the same batch. It wasn't
being reset once a request was sent, which prevented subsequent requests
from working.

Also, reduced the timeout from 500ms to 200ms. Waiting a half second
just doesn't feel right. I'd rather make a few separate requests in a
view with dozens of shortcodes than introduce extra latency into every
request.

goldenapples committed:
Use the counter as the object key in response

One step towards simplifying this data structure: rather than returning
the index as a attribute that has to be unwrapped to get at, use it as
the key in the ajax response object. (This also makes it easier to use
something other than a generic counter as the key if necessary.)

goldenapples committed:
Some exception handling

Add handling for a couple possible exceptions that could happen, if a
shortcode is deleted from the view between requesting its preview and
receiving the response.

goldenapples committed:
Trailing commas for phpcs validation

goldenapples committed:
Remove commented-out code

goldenapples committed:
Tabularize commas

danielbachhuber commented on inc/class-shortcode-ui.php:
```diff
@@ -338,8 +324,66 @@ public function handle_ajax_do_shortcode() {
+			exit;
+		}
+
+	}
```
Why do we need two separate methods? Can't we always just use one method which always accepts batch requests?

goldenapples replied:
I left the old method in there because I thought that it was handling something that core sent or expected. I guess not: removed in 87e7623.

danielbachhuber commented on js/src/utils/fetcher.js:
```diff
@@ -0,0 +1,75 @@
+
+		if ( ! fetcher.timeout ) {
+			fetcher.timeout = setTimeout(
+				fetcher.fetchAll, 200
```
Is 200 going to be a magic number?

goldenapples committed:
Remove 200ms delay from setTimeout call

The delay wasn't necessary here - anything in a setTimeout call will
wait until the currently execulting script finishes, which is all that's
necessary here.

goldenapples committed:
Remove unused ajax handler for `do_shortcode`
